src/main/kotlin/de/bund/digitalservice/useid/config/SecurityConfig.kt

```diff
@@ -13,31 +13,32 @@ import org.springframework.security.web.server.authentication.ServerAuthenticati
 
 @Configuration
 @EnableWebFluxSecurity
-class SecurityConfig {
-
+class SecurityConfig(
+    private val cspProperties: ContentSecurityPolicyProperties,
+    private val authenticationManager: ReactiveAuthenticationManager,
+    private val authenticationConverter: ServerAuthenticationConverter
+) {
     @Bean
-    fun springSecurityFilterChain(
-        http: ServerHttpSecurity,
-        authenticationManager: ReactiveAuthenticationManager,
-        authenticationConverter: ServerAuthenticationConverter
-    ): SecurityWebFilterChain {
+    fun springSecurityFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {
         return http.authorizeExchange()
             .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/*/tc-token").permitAll()
             .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/**").authenticated()
             .anyExchange().permitAll()
             .and().csrf().disable()
-            .headers().frameOptions().disable()
+            .headers()
+            .contentSecurityPolicy(cspProperties.getCSPHeaderValue()).and()
```

Comment:
I assume this sets the CSP header for all endpoints. Don't we only need it for the `/widget` endpoint?

Reply:
what I did earlier, I created 2 type security chains: `apiSpringSecurityFilterChain` and `webSpringSecurityFilterChain`, so that we can control from Spring Security rather Controller specific, in case we want to send frame-ancestors in controllers that have view e.g. `/incompatible`, but this is somehow not working as expected
Do you know where I made the mistake?

```kotlin
    fun apiSpringSecurityFilterChain(
        http: ServerHttpSecurity,
        authenticationManager: ReactiveAuthenticationManager,
        authenticationConverter: ServerAuthenticationConverter
    ): SecurityWebFilterChain {
        return http.authorizeExchange()
            .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/*/tc-token").permitAll()
            .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/**").authenticated()
            .anyExchange().permitAll()
            .and().addFilterAfter(
                apiAuthenticationFilter(authenticationManager, authenticationConverter),
                SecurityWebFiltersOrder.REACTOR_CONTEXT
            )
            .build()
    }
```

```kotlin
fun webSpringSecurityFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {

        return http.authorizeExchange()
            .pathMatchers("/widget").permitAll()
            .anyExchange().permitAll()
            .and().csrf().disable()
            .headers()
            .contentSecurityPolicy(contentSecurityPolicy.toString())
            .and()
            .frameOptions().disable()
            .and()
            .build()
```

Reply:
You inject the same `ServerHttpSecurity` singleton in both cases, so I guess one config overrides the other. And even if it just adds the config instead of overriding it, it would be applied to the same security chain so the headers would be applied to all endpoints. What you need is a config on controller or `pathMatcher` level.

Reply:
ok I guess for now let's do it more specific in the `/widget` controller